services/web-server/src/PulseEngine/PulseIterator.js

```diff
@@ -38,40 +44,65 @@ export default class PulseIterator {
     return this;
   }
 
+  // cause this to throw the given error for all pending and subsequent next() calls
+  pushError(error) {
+    this.stop({error, done: true});
+  }
+
+  // insert a value into the queue; the resulting promise resolves when the value
+  // is consumed, and rejects if the iterator finishes without consuming the value
   pushValue(value) {
-    if (this.pullQueue.size !== 0) {
-      this.pullQueue.first()({ value, done: false });
-      this.pullQueue = this.pullQueue.shift();
-    } else {
-      this.pushQueue = this.pushQueue.push(value);
```

Comment:
Interesting, in this case `next()` would not return a promise and would not indicate whether we're done or not. Was it causing any troubles, just out of curiosity?

Reply:
This is the case where pullQueue is empty, so there are no pending `next` calls waiting to return anything.

Reply:
why are there two "queues"? I don't understand this whole design

Reply:
The class matches up calls to `next()` and calls to `pushValue()`, either of which can happen at any time.  It does so by keeping a queue of each (`pullQueue` and `pushQueue`, respectively) and matching off items from those queues until one is empty (in the `match` method).